packages/loopring_v3/contracts/impl/libtransactions/WithdrawTransaction.sol

```diff
@@ -0,0 +1,229 @@
+/// @author Brecht Devos - <brecht@loopring.org>
+/// @dev The following 4 types of withdrawals are supported:
+///      - withdrawType = 0: offchain withdrawals with EdDSA signatures
+///      - withdrawType = 1: offchain withdrawals with ECDSA signatures or onchain appprovals
+///      - withdrawType = 2: onchain valid forced withdrawals (owner and accountID match), or
+///                          offchain operator-initiated withdrawals for protocol fees or for
+///                          users in shutdown mode
+///      - withdrawType = 3: onchain invalid forced withdrawals (owner and accountID mismatch)
+library WithdrawTransaction
+{
+    using BytesUtil            for bytes;
+    using FloatUtil            for uint;
+    using MathUint             for uint;
+    using SignatureUtil        for bytes32;
+    using ExchangeMode         for ExchangeData.State;
+    using ExchangeWithdrawals  for ExchangeData.State;
+
+    bytes32 constant public WITHDRAWAL_TYPEHASH = keccak256(
+        "Withdrawal(address owner,uint24 accountID,uint32 nonce,uint16 tokenID,uint256 amount,uint16 feeTokenID,uint256 fee,address to,bytes32 dataHash,uint24 minGas)"
+    );
+
+    struct Withdrawal
+    {
+        uint    withdrawalType;
+        address owner;
+        uint24  accountID;
+        uint32  nonce;
+        uint16  tokenID;
+        uint    amount;
+        uint16  feeTokenID;
+        uint    fee;
+        address to;
+        bytes32 dataHash;
+        uint24  minGas;
+    }
+
+    // Auxiliary data for each withdrawal
+    struct WithdrawalAuxiliaryData
+    {
+        uint  gasLimit;
+        bytes signature;
+        bytes auxiliaryData;
+    }
+
+    event ForcedWithdrawalProcessed(
+        uint24 accountID,
+        uint16 tokenID,
+        uint   amount
+    );
+
+    function process(
+        ExchangeData.State        storage S,
+        ExchangeData.BlockContext memory  ctx,
+        bytes                     memory  data,
+        bytes                     memory  auxiliaryData
+        )
+        internal
+        returns (uint feeETH)
+    {
+        Withdrawal memory withdrawal = readWithdrawal(data);
+        WithdrawalAuxiliaryData memory auxData = abi.decode(auxiliaryData, (WithdrawalAuxiliaryData));
+
+        // Validate gas provided
+        require(auxData.gasLimit >= withdrawal.minGas, "OUT_OF_GASH_FOR_WITHDRAWAL");
+
+        // Validate the auxixliary withdrawal data
+        if (withdrawal.dataHash == 0) {
+            require(auxData.auxiliaryData.length == 0, "AUXILIARY_DATA_NOT_ALLOWED");
+        } else {
+            // Hashes are stored using only 253 bits so the value fits inside a SNARK field element.
+            require(
+                uint(keccak256(auxData.auxiliaryData)) >> 3 == uint(withdrawal.dataHash),
+                "INVALID_WITHDRAWAL_AUX_DATA"
+            );
+        }
+
+        if (withdrawal.withdrawalType == 0) {
+            // Signature checked offchain, nothing to do
```

Comment:
I think for off chain EdDSA withdraw, we will never reach this condition branch because for such case, we won't put any data for conditional transactions. So here no need to judge equal to 0, just fallback to revert branch is enough.

Reply:
A withdrawal is always seen as a conditional transaction because it needs to be processed onchain (we actually have to do the token transfer or make it the amount withdrawn available onchain), so EdDSA/ECDSA/... doesn't matter specifically for withdrawals. For simple transfers you'd be absolutely right because the transfer itself is fully on layer 2 and would no onchain work at all.